### PR TITLE
Add HWRPs for installing and managing consul.

### PR DESCRIPTION
**Pull request by johnbellone**
I am channeling @someara and @reset here.

After taking a look at the new [httpd library cookbook][1] as well
as the [Mysql library cookbook][2] I think this is an excellent example
of using HWRP for defining the building blocks of a quality cookbook.

[1]: https://github.com/chef-cookbooks/httpd
[2]: https://github.com/chef-cookbooks/mysql

### CODE REVIEW

**Comment by reset** [comment]
@johnbellone excellent work man! I really like having the ability to swap service providers for the service resource. That'll let people implement and test whatever service startup tool they use (runit, systemd, etc). Comment here again when it's finished and I'll take another look ;)

File: libraries/provider_consul_client_source.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class Chef::Provider::ConsulServiceSource < Chef::Provider::ConsulService
+  action :create do
+    include_recipe 'golang::default'
+
+    directory File.join(node['go']['gopath'], 'src/github.com/hashicorp') do
```
Comment (someara) [inline libraries/provider_consul_client_source.rb]: It's not a good idea to consume node attributes inside a provider... pass this in through the resource interface. You can do type checking (kind_of: String) there as well. I'd also use a method to calculate the string.

File: libraries/provider_consul_client_source.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      owner 'root'
+      group 'root'
+      mode '00755'
+      recursive true
+      action :create
+    end
+
+    git File.join(node['go']['gopath'], '/src/github.com/hashicorp/consul') do
```
Comment (someara) [inline libraries/provider_consul_client_source.rb]: node attributes, mkay

**Comment by someara** [comment]
This is great. I'm so happy to see people copying this pattern.

+100.

**Comment by fujin** [comment]
hey there, tried to give this a test, but getting:

```
* consul_client[/opt/consul/0.4.1] action create (up to date)
       
           
           ================================================================================
           Error executing action `create` on resource 'consul_config[/etc/consul.d/default.json]'
           ================================================================================
           
           NoMethodError
           -------------
           undefined method `parsed_run_user' for Chef::Resource::Directory
           
           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/consul/libraries/provider_consul_config.rb:20:in `block (2 levels) in <class:ConsulConfig>'
           /tmp/kitchen/cache/cookbooks/consul/libraries/provider_consul_config.rb:19:in `block in <class:ConsulConfig>'
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/consul/recipes/default.rb
           
            18: config = consul_config File.join(node['consul']['config_dir'], 'default.json') do
            19:   run_user client.run_user
            20:   run_group client.run_group
            21: end
            22: 
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/consul/recipes/default.rb:18:in `from_file'
           
           consul_config("/etc/consul.d/default.json") do
             action :create
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :consul_config
             cookbook_name "consul"
             recipe_name "default"
             run_user "consul"
             run_group "consul"
             path "/etc/consul.d/default.json"
           end
           
       [2015-03-09T20:46:36+00:00] ERROR: Converge failed with error message consul_config[/etc/consul.d/default.json] (consul::default line 18) had an error: NoMethodError: undefined method `parsed_run_user' for Chef::Resource::Directory
```

I looked into it and it would appear that the directory resource being used in consul_config is trying to use a class method from ConsulCookbook::Helpers, parsed_run_user. I see in some previous commits there was a change from run_user to service_user, and I wonder if this has just not been updated yet.

directory: https://github.com/johnbellone/consul-cookbook/blob/develop/libraries/provider_consul_config.rb#L20-L21
helper methods: https://github.com/johnbellone/consul-cookbook/blob/develop/libraries/helpers.rb#L13-L21

thanks in advance for any thoughts, let me know if I can help in anyway.

**Comment by johnbellone** [comment]
@fujin I have a whole bunch of commits I need to push. Apologies. They'll be coming in a few minutes.

**Comment by johnbellone** [comment]
I need to tear into this now that I am back from vacation. I'll see if I get some time to hack on it this evening.

**Comment by johnbellone** [comment]
This has become a beastly set of changes, but I am nearing completion. I need to open source a cookbook from work prior to release and thorough testing.

**Comment by johnbellone** [comment]
/cc @reset @someara 

This is ready for review and merging.

**Comment by johnbellone** [comment]
There's still a lot of testing left. But I'd like to get it into master and have community help out :).

File: README.md [hunk 3]
```diff
@@ -1,491 +1,50 @@
...
-Installing the separate Consul UI simply requires you to include
-the `consul::ui` recipe in your node's `run_list`.
-
-```json
-{
-  "run_list": [
-    "recipe[consul::ui]"
-  ]
-}
+### consul_definition
```
Comment (reset) [inline README.md]: Need to fill out an example here

File: README.md [hunk 4]
```diff
@@ -1,491 +1,50 @@
...
+### consul_service
+```ruby
+consul_service 'consul' do
+  user 'consul'
+  group 'consul'
+  install_method :binary
+  binary_url node['consul']['binary_url']
+end
 ```
+### consul_watch
```
Comment (reset) [inline README.md]: Need to fill out an example here

File: README.md [hunk 5]
```diff
@@ -1,491 +1,50 @@
...
-the `consul::ui` recipe in your node's `run_list`.
-
-```json
-{
-  "run_list": [
-    "recipe[consul::ui]"
-  ]
-}
+### consul_definition
+### consul_service
```
Comment (reset) [inline README.md]: I think this should be the first documented LWRP and immediately follow a quick "usage" section. Usage should include instructions for including the default recipe and explain what you get and also how to leverage the consul-service lwrp

**Comment by reset** [comment]
Huge sweeping changes so it's difficult to give a proper code review, but this looks good on the surface. I included a few comments on the README. I'm a bit concerned with the addition of poise and chef vault as dependencies. Why did we pick both of these up?

**Comment by johnbellone** [comment]
The usage of chef-vault is to seed initial secrets and SSL certificates. The poise usage is specifically for poise-service which abstracts away all of the sysv, upstart, systemd and runit cruft.

**Comment by reset** [comment]
@johnbellone makes sense to me!

**Comment by johnbellone** [comment]
You can take a look at [vault-cluster cookbook](https://github.com/johnbellone/vault-cluster-cookbook) and [hashicorp-vault cookbook](https://github.com/johnbellone/vault-cookbook) which utilize patterns. The former uses the development branch here to standup Consul/Vault on same instance.

**Comment by johnbellone** [comment]
I'll write some documentation and merge this in.

**Pull request merged**